astropy/utils/misc.py

```diff
@@ -850,3 +850,121 @@ def set_locale(name):
                 yield
             finally:
                 locale.setlocale(locale.LC_ALL, saved)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class ShapedLikeNDArray(object):
+    """Mixin class to provide shape-changing methods.
+
+    The class proper must define abstract properties ``ndim`` and ``shape``,
+    as well as an abstract method ``_replicate``.
+    """
+
+    @abc.abstractproperty
+    def shape(self):
+        """The shape of the instance and underlying arrays."""
+
```

Comment:
I realised that for `ndim` one also has a sensible default: `len(self.shape)`. So I added that as well.